### PR TITLE
Event payload lives in the LogRecord body

### PR DESCRIPTION
**Pull request by breedx-splk**
## Changes

In the Events Working Group on 2023-11-17, we decided that the Event payload will exist in the `Body` of the `LogRecord`. This provides that clarity and also provides some additional descriptive wording about what an event is.

This also relates to https://github.com/open-telemetry/opentelemetry-specification/pull/3772

## Merge requirement checklist

* [ ] [CONTRIBUTING.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md) guidelines followed.
* [ ] [CHANGELOG.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CHANGELOG.md) updated for non-trivial changes.
* [ ] [schema-next.yaml](https://github.com/open-telemetry/semantic-conventions/blob/main/schema-next.yaml) updated with changes to existing conventions.

### CODE REVIEW

File: docs/general/events.md [hunk 1]
```diff
@@ -3,13 +3,34 @@ linkTitle: Events
...
+Examples of Events might include things like uncaught exceptions, button clicks, user logout,
+network connection severed, etc.
+
+In OpenTelemetry, Events are implemented as a specific type of `LogRecord` that conforms to
+the conventions included here, and Events
+[have their own api](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/event-api.md).
+The api abstracts away knowledge of `LogRecord` so that users are merely dealing with Event
+semantics.
+
+In addition to a required name, an Event may contain a _payload_, which can be either structured or unstructured.
```
[the remark this answers is not in the document]
Comment (breedx-splk) [reply docs/general/events.md:24]: > We require the reader to make the logical leap to go and find the [AnyValue](https://github.com/open-telemetry/opentelemetry-proto/blob/ea449ae0e9b282f96ec12a09e796dbb3d390ed4f/opentelemetry/proto/common/v1/common.proto#L28) definition in the proto, which is asking a lot.

Yeah, that's kinda why I was trying to avoid it because hopefully it's abstracted away. What about something like this:

"In addition to a required name, an Event may contain a _payload_ of any type permitted by the [`LogRecord` body](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#field-body). "

[the remark this answers is not in the document]
Comment (scheler) [reply docs/general/events.md:24]: Should we add that Events also have a priority, in addition to name, optional payload and optional attributes? The Event API talks about priority, so it will help with being consistent with it if we add priority to the definition or description of what Events contain.

File: docs/general/events.md [hunk 2]
```diff
@@ -3,13 +3,34 @@ linkTitle: Events
...
+In its implementation, the Event _payload_ will constitute the `Body` of the `LogRecord`.
+Like all other OpenTelemetry signals, an Event has optional attribute metadata that helps describe
+the event context.
+
+Over time, some Events will be specified by OpenTelemetry and will have documented payload structure,
+field semantics, and stability and requirement levels. Other events may be user-defined and carry
+bespoke user semantics. The name of the Event determines if an established payload structure is applicable
+or not.
+
```
Comment (scheler) [inline docs/general/events.md:33]: Can you clarify what you mean in the last sentence? By established payload structure, do you mean schema of the event? I am of the understanding that any event names documented in this repo will also have its payload structure documented, so it's not clear what you mean by a schema not being application for certain event names.

Comment (breedx-splk) [reply docs/general/events.md:33]: > By established payload structure, do you mean schema of the event?

Yeah, a schema is what would be used to describe the structure in most cases. If you think that using the word schema here helps, we can try it out. I mostly didn't want to get too specific about a thing that doesn't exist yet.

> I am of the understanding that any event names documented in this repo will also have its payload structure documented, so it's not clear what you mean by a schema not being applicable for certain event names.

The intent, without getting too deep into the event registry and schemas (both of which are germinating), was to suggest that the name of the event determines if a predefined structure applies. That is, an event like `otel.button.click` should be in the registry within the spec and have a defined structure (schema), but `jimmy.door.knock` might not.

**Comment by breedx-splk** [comment]
>Should we add that Events also have a priority, in addition to name, optional payload and optional attributes? The Event API talks about priority, so it will help with being consistent with it if we add priority to the definition or description of what Events contain.

Maybe, but if so probably in a follow-up PR. The intent of this PR was the help clarify that the payload of an event lives in the body of the `LogRecord`.

File: docs/general/events.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (breedx-splk) [reply docs/general/events.md:24]: Maybe, but if so probably in a follow-up PR. The intent of this PR was the help clarify that the payload of an event lives in the body of the LogRecord.

File: docs/general/events.md (hunk 2 above)
Comment (scheler) [reply docs/general/events.md:34]: I am still unclear, if you are suggesting a convention for event names for events in the semconv registry, like `otel` prefix or just that if an event name is defined in the registry it also has its payload structure defined. If it's the latter, I suggest rewording to something like this - "All events defined in the semantic conventions include both the event name and its payload structure"..

This brings up a question whether we allow defining event names  in semantic conventions for external events, like Kubernetes events, for which we cannot defining the structure since it's not in our control.

Comment (breedx-splk) [reply docs/general/events.md:34]: > if you are suggesting a convention for event names for events in the semconv registry

No, this would not be the right PR for that.

> if an event name is defined in the registry it also has its payload structure defined

Basically. The intent is more like "If you get an event, how do you know if there's a well known structure for the event payload? You look at the name. The body itself doesn't tell if there is a schema, the name is what determines it."

But I think I like your idea `"All events defined in the semantic conventions include both the event name and its payload structure"`. Let me try something more like that.

Comment (breedx-splk) [reply docs/general/events.md:34]: Ok, try out that new wording.

**Review by scheler — APPROVED** [review]

File: docs/general/events.md [hunk 3]
```diff
@@ -3,13 +3,35 @@ linkTitle: Events
 aliases: [docs/specs/semconv/general/events-general]
 --->
 
-# Semantic Conventions for Event Attributes
+# Semantic Conventions for Events
 
 **Status**: [Experimental][DocumentStatus]
 
-This document describes the attributes of standalone Events that are represented
+This document describes the characteristics of standalone Events that are represented
```
Comment (jsuereth) [inline docs/general/events.md:10]: Remember that we use "Event" to also means "SpanEvent" - https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#add-events

I'm not sure you can just update this section to only be about LogRecord without also touching/describing SpanEvents (which is what this section was initially created to deal with).

Comment (breedx-splk) [reply docs/general/events.md:10]: The intent was not to just describe `LogRecord` as you suggested, but rather to start to flesh out Events as their own first-class signal type. The fact that it they are built on top of or transported via logs should be a separate (implementation) concern. The specific line you commented on changes "attributes" to "characteristic" because Attributes has lots of existing meaning on otel and I didn't want people to think that "attributes" meant "Attributes" while that decision is still being worked on by the events sig.

I don't think we have indicated that Events and SpanEvents are the same thing. @jsuereth are you suggesting that we have? Or maybe we want to?

Comment (trask) [reply docs/general/events.md:10]: > I'm not sure you can just update this section to only be about LogRecord without also touching/describing SpanEvents (which is what this section was initially created to deal with).

my understanding is that this page only deals with the new experimental (log-based) Events, while Span Events are described here: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#add-events

**Review by trask — APPROVED** [review]

**Comment by trask** [comment]
Adding reason for my approval:

Putting the payload in the LogRecord body makes a clear delineation between
* attributes (metadata), and
* body (payload)

(ok, it's still a bit messy whether some fields should go in payload or metadata, but I think that discussion can happen when we start adding semantic conventions for specific events).

Further (and importantly), we can then restrict the Event API to only support non-nested attributes, which allows us to bypass the can of worms in https://github.com/open-telemetry/opentelemetry-specification/pull/2888

File: docs/general/events.md (hunk 3 above)
Comment (jsuereth) [reply docs/general/events.md:10]: My concern is that, reading this, I don't see how it can apply to today's `SpanEvent`.  I like the content you've written and I would otherwise approve this, but I think this will only continue to confuse folks figuring out `SpanEvent` vs. `Event`.

Example confusing language:

```
In addition to a required name, an Event may contain a _payload_ of any type permitted by the
[LogRecord body](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#field-body).
In its implementation, the Event _payload_ will constitute the `Body` of the `LogRecord`.
Like all other OpenTelemetry signals, an Event has optional attribute metadata that helps describe
the event context.
```

How does that interact with this: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#add-events

**Comment by yurishkuro** [comment]
Shouldn't this be in specification repo? semconv do not define APIs, and where data goes when passed to the API is a spec issue.

**Review by tedsuo — APPROVED** [review]

File: docs/general/events.md [hunk 4]
```diff
@@ -3,13 +3,35 @@ linkTitle: Events
...
+In addition to a required name, an Event may contain a _payload_ of any type permitted by the
+[LogRecord body](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#field-body).
+In its implementation, the Event _payload_ will constitute the `Body` of the `LogRecord`.
+Like all other OpenTelemetry signals, an Event has optional attribute metadata that helps describe
+the event context.
+
+Over time, some Events will be specified by OpenTelemetry and will have documented payload structure,
+field semantics, and stability and requirement levels. Other events may be user-defined and carry
+bespoke user semantics. When an Event name exists in the semantic conventions, its _payload_
+structure and semantics will also be defined.
```
Comment (lmolkova) [inline docs/general/events.md:33]: I really want to see some specifics on how events can be described in semconv, specifically wrt payload/body.
It's not possible today and without defining a mechanism this PR seems incomplete.

Comment (breedx-splk) [reply docs/general/events.md:33]: Thanks @lmolkova. This is definitely intended to be incremental progress and not a comprehensive solve for events. We (the events working group) are working through some of the ideas around semconv and how it might play with existing tooling (or not). 

This PR was not intending to do all of that in one shot, but rather simply solidify the decision that the event payload lives in the log record body. 

How event payloads are eventually described in semantic conventions, how schemas might be represented, how markdown might be generated from yaml/schema files, how changes to existing events can be expressed in schema version files....all of that is still being worked out.

**Comment by yurishkuro** [comment]
The [Event API already says](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/event-api.md#emit-event) that payload is placed in the body. It also says event may have attributes. Neither the spec nor this PR specify what a "payload" is.

File: docs/general/events.md (hunk 4 above)
Comment (lmolkova) [reply docs/general/events.md:33]: Thank you for the explanation! 

So if I understand correctly, this PR only declares that payload should be described by the semconv and there will be a way. 
I'm glad there is a concensus there, but without specifics, what's the value of having this in the semconv?

**Comment by lmolkova** [comment]
I'd like to see a story for span events. It can look as simple as:
- define `event.payload` attribute (for `event` type which is spans event) - string.
- define other similar attributes for properties that Event has but span event doesn't  (seem to be `SeverityNumber` only)

Then semconv authors would be able to describe events using schemas/mds the moment this PR is merged.
Exporters could start translating span events to events if they want to.

Events WG can keep working on defining the future of Event/Log semconv in parallel.

File: docs/general/events.md (hunk 4 above)
Comment (breedx-splk) [reply docs/general/events.md:33]: Previously, there was an open question on where the event payload would live -- does it go in the body of the LogRecord, or does it go into an attribute with a specific name like `event.data`? The benefit is having this question answered and providing a foundation on which to build future semconv.

**Comment by breedx-splk** [comment]
> I'd like to see a story for span events. <snip>

Thanks @lmolkova. A while back, @jsuereth also brought up span events as a potential point of confusion.

Did you see [Trask's comment](https://github.com/open-telemetry/semantic-conventions/pull/566/files#r1464047469) up there? This page is intended only to describe stand-alone events, and makes no reference to Spans anywhere. The only reasonable way I could understand a user being confused between events and span events is if they came from a tracing background and just ignored the fact that span events have the word span in them. :)

Do we really think this is a blocker for _this_ PR, or is it something we can work on after to unify/clarify the relationship between events and span events in follow-up work?

File: docs/general/events.md [hunk 5]
```diff
@@ -3,13 +3,35 @@ linkTitle: Events
...
+Examples of Events might include things like uncaught exceptions, button clicks, user logout,
+network connection severed, etc.
+
+In OpenTelemetry, Events are implemented as a specific type of `LogRecord` that conforms to
+the conventions included here, and Events
+[have their own API](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/event-api.md).
+The API abstracts away knowledge of `LogRecord` so that users only deal with Event
+semantics.
+
+In addition to a required name, an Event may contain a _payload_ of any type permitted by the
```
Comment (yurishkuro) [inline docs/general/events.md:24]: Can we define what a "payload" is semantically? Is it the actual structural representation of the event, or an arbitrary "attachment" (a custom user-defined structure that can vary for the same event type)?

File: docs/general/events.md (hunk 3 above)
Comment (jsuereth) [reply docs/general/events.md:10]: So regarding this concern.  I think we can take the Event/Span-Event confusion into a separate PR or even issue that we track to clean up the language over semconv.  I'd like to make progress here, and I think calling out that log-body has its own fields that are not shared with "Attributes" is important.


Could you open a ticket where we can evaluate and track confusing aspects of the specification going forward regarding "Events (Logs)" vs. "(Span)Events" ?

**Review by jsuereth — APPROVED** [review]

**Review by martinkuba — APPROVED** [review]

**Review by scheler — APPROVED** [review]

File: docs/general/events.md (hunk 5 above)
Comment (breedx-splk) [reply docs/general/events.md:24]: Yes, I think this will happen as follow-up work. I think defining details of the "payload" can be handled orthogonally from where payload lives, which is what this PR is about.

**Review by arminru — APPROVED** [review]

File: docs/general/events.md (hunk 5 above)
Comment (yurishkuro) [reply docs/general/events.md:24]: I am not sure I follow this reasoning - you want to add something to the spec without defining what it is? What difference does it make _where_ payload lives if "payload" is undefined?

Comment (MSNev) [inline docs/general/events.md:24]: Well actually been defining that the _payload_ is a Map<string, Attribute> values rather than `of any type` as the Event Api is being defined as receiving Attributes whichit will then place into the `body`. Likewise with the prototyping of the build-tools...

**Review by joaopgrassi — APPROVED** [review]

File: docs/general/events.md (hunk 5 above)
Comment (breedx-splk) [reply docs/general/events.md:24]: That's fine. I'm not sure that's incompatible with the vernacular here, is it? 

This was trying to just say that the payload is whatever is allowed by the body. If you brought this up because you think it needs to be more specific here, can we do that as a more precise follow-up PR?

**Comment by breedx-splk** [comment]
Adding purely informative link to [related spec PR](https://github.com/open-telemetry/opentelemetry-specification/pull/3858).

File: docs/general/events.md (hunk 3 above)
Comment (breedx-splk) [reply docs/general/events.md:10]: Link to [#695](https://github.com/open-telemetry/semantic-conventions/issues/695).

File: docs/general/events.md (hunk 4 above)
Comment (lmolkova) [reply docs/general/events.md:33]: I'm going to resolve it in favor of https://github.com/open-telemetry/semantic-conventions/issues/695

File: docs/general/events.md (hunk 5 above)
Comment (jsuereth) [reply docs/general/events.md:24]: We are planning to add payloads into semconv, including rules of what they are and ways to denote their schema going forward.

This is just step one for clarity.
